File: python/test/t_JunctionTreeBernsteinCopulaFactory_std.py
#!/usr/bin/env python

import openturns as ot
import otagrum


def generateData(size):
    dim = 4
    R = ot.CorrelationMatrix(dim)
    for i in range(dim):
        for j in range(i):
            R[i, j] = 0.3
    ot.RandomGenerator.SetSeed(0)
    return ot.Normal([0.0] * dim, R).getSample(size)


def testBuildFromSample():
    dim = 4
    sample = generateData(500)

    factory = otagrum.JunctionTreeBernsteinCopulaFactory(5, 0.1, 3)
    copula = factory.buildAsJunctionTreeBernsteinCopula(sample)

    print("learnt copula dimension : " + ("OK" if copula.getDimension() == dim else "fail"))

    generic = factory.build(sample)
    print(
        "build() dimension matches : "
        + ("OK" if generic.getDimension() == dim else "fail")
    )


def testEmptySampleRaises():
    factory = otagrum.JunctionTreeBernsteinCopulaFactory()
    try:
        factory.buildAsJunctionTreeBernsteinCopula(ot.Sample())
        print("empty sample raises : fail (no exception)")
    except Exception:
        print("empty sample raises : OK")


def testDefaultBuild():
    factory = otagrum.JunctionTreeBernsteinCopulaFactory()
    empty = factory.buildAsJunctionTreeBernsteinCopula()
    print("default build dimension : " + ("OK" if empty.getDimension() == 1 else "fail"))


testBuildFromSample()
testEmptySampleRaises()
testDefaultBuild()

File: lib/test/t_JunctionTreeBernsteinCopulaFactory_std.cxx
#include <cmath>
#include <iostream>

#include <openturns/OT.hxx>
#include "otagrum/otagrum.hxx"

using namespace OTAGRUM;

void testBuildFromSample()
{
  const OT::UnsignedInteger dim = 4;
  OT::CorrelationMatrix R(dim);
  for (OT::UnsignedInteger i = 0; i < dim; ++i)
    for (OT::UnsignedInteger j = 0; j < i; ++j)
      R(i, j) = 0.3;

  OT::RandomGenerator::SetSeed(0);
  OT::Sample sample(OT::Normal(OT::Point(dim), R).getSample(500));

  JunctionTreeBernsteinCopulaFactory factory(5, 0.1, 3);
  JunctionTreeBernsteinCopula copula = factory.buildAsJunctionTreeBernsteinCopula(sample);

  std::cout << "learnt copula dimension : "
            << ((copula.getDimension() == dim) ? "OK" : "fail") << std::endl;

  OT::RandomGenerator::SetSeed(33);
  const double entropy =
    -copula.computeLogPDF(copula.getSample(500)).computeMean()[0];
  std::cout << "entropy is finite : "
            << (std::isfinite(entropy) ? "OK" : "fail") << std::endl;

  // build() (the OT::DistributionFactoryImplementation entry point) should
  // agree with buildAsJunctionTreeBernsteinCopula() on the dimension.
  OT::Distribution generic = factory.build(sample);
  std::cout << "build() dimension matches : "
            << ((generic.getDimension() == dim) ? "OK" : "fail") << std::endl;
}

void testEmptySampleRaises()
{
  JunctionTreeBernsteinCopulaFactory factory;
  std::cout << std::endl << "Empty sample should raise : ";
  try
  {
    factory.buildAsJunctionTreeBernsteinCopula(OT::Sample());
    std::cout << "fail (no exception)" << std::endl;
  }
  catch (const OT::InvalidArgumentException &)
  {
    std::cout << "OK" << std::endl;
  }
}

void testDefaultBuild()
{
  JunctionTreeBernsteinCopulaFactory factory;
  JunctionTreeBernsteinCopula empty = factory.buildAsJunctionTreeBernsteinCopula();
  std::cout << std::endl << "Default build() dimension : "
            << ((empty.getDimension() == 1) ? "OK" : "fail") << std::endl;
}

int main(int /*argc*/, char ** /*argv*/)
{
  testBuildFromSample();
  testEmptySampleRaises();
  testDefaultBuild();
  return EXIT_SUCCESS;
}
